tcmur_cmd_handler.c

```diff
@@ -181,6 +181,340 @@ struct write_same {
 	size_t iov_len;
 };
 
+struct unmap_state {
+	pthread_mutex_t lock;
+	unsigned int refcount;
+	bool error;
+};
+
+struct unmap_descriptor {
+	uint64_t offset;
+	uint64_t length;
+
+	struct tcmulib_cmd *origcmd;
+};
+
+static void handle_unmap_cbk(struct tcmu_device *dev, struct tcmulib_cmd *ucmd, int ret)
+{
+	struct unmap_descriptor *desc = ucmd->cmdstate;
+	struct tcmulib_cmd *origcmd = desc->origcmd;
+	struct unmap_state *state = origcmd->cmdstate;
+	bool error;
+
+	free(desc);
+	free(ucmd);
+
+	pthread_mutex_lock(&state->lock);
+	if (--state->refcount > 0) {
+		pthread_mutex_unlock(&state->lock);
+		return;
+	}
+	error = state->error;
```

Comment:
Sorry, one more comment.

There is a bug in my patch in PR 243 and here.

If the handler->unmap call fails, then it copies sense to the ucmd->sense_buf. In this case we will need to copy that sense to origcmd->sense_buf.